### ISSUE TITLE
Datarace during tests execution

### ISSUE THREAD

**Issue by ialidzhikov** [gardener/dependency-watchdog#12]
The issue occurred during tests execution for #11. 

See build https://concourse.ci.infra.gardener.cloud/builds/80565.

See the output below.

```
+ git config -f pull-request-gardener_dependency-watchdog-pr_master/.git/config pullrequest.id
+ export PULLREQUEST_ID=11
+ git config -f pull-request-gardener_dependency-watchdog-pr_master/.git/config pullrequest.url
+ export PULLREQUEST_URL=https://github.com/gardener/dependency-watchdog/pull/11
+ cat managed-version/version
+ export EFFECTIVE_VERSION=v0.4.1-dev-836f43b1813589980fe4aebb096e4307dd3174b6
+ cp managed-version/version pull-request-gardener_dependency-watchdog-pr_master/VERSION
+ readlink -f .
+ readlink -f .
+ CC_ROOT_DIR=/tmp/build/a94a8fe5
+ export CC_ROOT_DIR
+ [ -x pull-request-gardener_dependency-watchdog-pr_master/.ci/test ]
+ pull-request-gardener_dependency-watchdog-pr_master/.ci/test
go: finding github.com/gogo/protobuf v1.3.1
go: finding github.com/hashicorp/golang-lru v0.5.3
go: finding github.com/json-iterator/go v1.1.8
go: finding github.com/pkg/errors v0.8.1
go: finding github.com/evanphx/json-patch v4.5.0+incompatible
go: finding github.com/imdario/mergo v0.3.8
go: finding github.com/ghodss/yaml v1.0.0
go: finding golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
go: finding k8s.io/utils v0.0.0-20191114200735-6ca3b61696b6
go: finding k8s.io/component-base v0.0.0-20190516230545-9b07ebd4193b
go: finding github.com/googleapis/gnostic v0.3.1
go: finding github.com/onsi/ginkgo v1.10.3
go: finding github.com/hpcloud/tail v1.0.0
go: finding github.com/stretchr/testify v1.4.0
go: finding k8s.io/api v0.0.0-20190516230258-a675ac48af67
go: finding k8s.io/klog v1.0.0
go: finding github.com/pborman/uuid v1.2.0
go: finding k8s.io/client-go v11.0.0+incompatible
go: finding k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
go: finding golang.org/x/net v0.0.0-20191119073136-fc4aabc6c914
go: finding k8s.io/klog v0.3.0
go: finding github.com/spf13/cobra v0.0.5
go: finding k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
go: finding golang.org/x/sys v0.0.0-20191128015809-6d18c012aee9
go: finding gopkg.in/fsnotify.v1 v1.4.7
go: finding github.com/mailru/easyjson v0.0.0-20160728113105-d5b7844b561a
go: finding golang.org/x/text v0.0.0-20160726164857-2910a502d2bf
go: finding github.com/pmezard/go-difflib v0.0.0-20151028094244-d8ed2627bdf0
go: finding golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4
go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: finding gopkg.in/yaml.v2 v2.2.2
go: finding github.com/docopt/docopt-go v0.0.0-20180111231733-ee0de3bc6815
go: finding github.com/stretchr/testify v1.3.0
go: finding github.com/onsi/ginkgo v0.0.0-20170829012221-11459a886d9c
go: finding github.com/kisielk/gotool v1.0.0
go: finding github.com/kisielk/errcheck v1.2.0
go: finding github.com/mitchellh/go-homedir v1.1.0
go: finding github.com/json-iterator/go v0.0.0-20180612202835-f2b4162afba3
go: finding github.com/spf13/pflag v0.0.0-20170130214245-9ff6c6923cff
go: finding golang.org/x/time v0.0.0-20191024005414-555d28b269f0
go: finding github.com/modern-go/reflect2 v0.0.0-20180320133207-05fbef0ca5da
go: finding github.com/google/gofuzz v1.0.0
go: finding github.com/spf13/afero v1.2.2
go: finding github.com/googleapis/gnostic v0.0.0-20170729233727-0c5108395e2d
go: finding github.com/munnerz/goautoneg v0.0.0-20120707110453-a547fc61f48d
go: finding sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e
go: finding k8s.io/gengo v0.0.0-20190128074634-0689ccc1d7d6
go: finding github.com/golang/groupcache v0.0.0-20191027212112-611e8accdfc9
go: finding github.com/PuerkitoBio/purell v1.0.0
go: finding gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
go: finding github.com/onsi/gomega v0.0.0-20170829124025-dcabb60a477c
go: finding github.com/davecgh/go-spew v1.1.0
go: finding github.com/pmezard/go-difflib v1.0.0
go: finding golang.org/x/text v0.3.0
go: finding github.com/davecgh/go-spew v0.0.0-20151105211317-5215b55f46b2
go: finding google.golang.org/appengine v1.4.0
go: finding github.com/google/uuid v1.0.0
go: finding github.com/inconshreveable/mousetrap v1.0.0
go: finding github.com/stretchr/objx v0.1.0
go: finding github.com/cpuguy83/go-md2man v1.0.10
go: finding github.com/go-openapi/jsonreference v0.0.0-20160704190145-13c6e3589ad9
go: finding github.com/golang/protobuf v1.3.2
go: finding golang.org/x/tools v0.0.0-20181030221726-6c7e314b6563
go: finding github.com/modern-go/reflect2 v0.0.0-20180701023420-4b7aa43c6742
go: finding github.com/spf13/pflag v1.0.5
go: finding gopkg.in/yaml.v2 v2.2.1
go: finding github.com/go-openapi/spec v0.0.0-20160808142527-6aced65f8501
go: finding github.com/ghodss/yaml v0.0.0-20150909031657-73d445a93680
go: finding golang.org/x/tools v0.0.0-20181011042414-1f849cf54d09
go: finding golang.org/x/sys v0.0.0-20170830134202-bb24a47a89ea
go: finding github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46
go: finding github.com/davecgh/go-spew v1.1.1
go: finding github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: finding gopkg.in/inf.v0 v0.9.1
go: finding github.com/golang/protobuf v0.0.0-20161109072736-4bd1920723d7
go: finding golang.org/x/net v0.0.0-20170114055629-f2499483f923
go: finding github.com/modern-go/concurrent v0.0.0-20180228061459-e0a39a4cb421
go: finding sigs.k8s.io/yaml v1.1.0
go: finding github.com/stretchr/testify v0.0.0-20151208002404-e3a8ff8ce365
go: finding github.com/go-openapi/jsonpointer v0.0.0-20160704185906-46af16f9f7b1
go: finding github.com/emicklei/go-restful v0.0.0-20170410110728-ff4f55a20633
go: finding github.com/go-logr/logr v0.1.0
go: finding github.com/spf13/viper v1.3.2
go: finding golang.org/x/net v0.0.0-20180724234803-3673e40ba225
go: finding github.com/russross/blackfriday v1.5.2
go: finding github.com/golang/protobuf v1.2.0
go: finding golang.org/x/net v0.0.0-20190108225652-1e06a53dbb7e
go: finding cloud.google.com/go v0.34.0
go: finding github.com/modern-go/reflect2 v1.0.1
go: finding golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a
go: finding github.com/spf13/pflag v1.0.3
go: finding github.com/google/gofuzz v0.0.0-20161122191042-44d81051d367
go: finding github.com/go-openapi/swag v0.0.0-20160704191624-1d0bd113de87
go: finding golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2
go: finding github.com/spf13/afero v1.1.2
go: finding github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77
go: finding github.com/BurntSushi/toml v0.3.1
go: finding github.com/fsnotify/fsnotify v1.4.7
go: finding github.com/coreos/go-semver v0.2.0
go: finding github.com/magiconair/properties v1.8.0
go: finding github.com/coreos/go-etcd v2.0.0+incompatible
go: finding github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6
go: finding golang.org/x/crypto v0.0.0-20181203042331-505ab145d0a9
go: finding github.com/stretchr/testify v1.2.2
go: finding github.com/coreos/etcd v3.3.10+incompatible
go: finding github.com/PuerkitoBio/urlesc v0.0.0-20160726150825-5bd2802263f2
go: finding github.com/hashicorp/hcl v1.0.0
go: finding github.com/spf13/jwalterweatherman v1.0.0
go: finding golang.org/x/sys v0.0.0-20181205085412-a5c9d58dba9a
go: finding github.com/pelletier/go-toml v1.2.0
go: finding github.com/ugorji/go/codec v0.0.0-20181204163529-d75b2dcb6bc8
go: finding github.com/spf13/cast v1.3.0
go: finding k8s.io/klog v0.0.0-20181102134211-b9b56d5dfc92
go: finding github.com/mitchellh/mapstructure v1.1.2
go: finding github.com/onsi/ginkgo/ginkgo latest
go: finding github.com/onsi/ginkgo v1.12.0
go: downloading github.com/onsi/ginkgo v1.12.0
go: extracting github.com/onsi/ginkgo v1.12.0
go: finding github.com/onsi/gomega v1.7.1
go: finding golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e
go: finding golang.org/x/sys latest
go: finding github.com/onsi/gomega v1.9.0
go: finding golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f
go: finding github.com/onsi/ginkgo v1.6.0
go: finding gopkg.in/yaml.v2 v2.2.4
go: finding gopkg.in/tomb.v1 latest
go: finding golang.org/x/text v0.3.2
go: finding golang.org/x/net v0.0.0-20180906233101-161cd47e91fd
go: finding golang.org/x/sys v0.0.0-20180909124046-d0be0721c37e
go: finding github.com/golang/protobuf v1.3.3
go: finding golang.org/x/sync latest
go: finding golang.org/x/net latest
go: finding golang.org/x/tools v0.0.0-20180917221912-90fa682c2a6e
go: finding golang.org/x/tools latest
go: finding golang.org/x/crypto latest
go: finding golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
go: finding golang.org/x/sys v0.0.0-20190412213103-97732733099d
go: finding golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: finding golang.org/x/net v0.0.0-20190620200207-3b0461eec859
go: finding golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7
go: finding golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898
go: finding golang.org/x/mod v0.1.1-0.20191105210325-c90efee705ee
go: finding golang.org/x/xerrors latest
go: finding golang.org/x/mod v0.2.0
go: finding golang.org/x/crypto v0.0.0-20191011191535-87dc89f01550
go: finding golang.org/x/tools v0.0.0-20191119224855-298f0cb1881e
go: finding gopkg.in/yaml.v2 v2.2.8
go: finding gopkg.in/check.v1 latest
go: downloading github.com/hpcloud/tail v1.0.0
go: extracting github.com/hpcloud/tail v1.0.0
go: downloading gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: downloading gopkg.in/fsnotify.v1 v1.4.7
go: extracting gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: extracting gopkg.in/fsnotify.v1 v1.4.7
go: downloading golang.org/x/sys v0.0.0-20200202164722-d101bd2416d5
go: extracting golang.org/x/sys v0.0.0-20200202164722-d101bd2416d5
[INFO] Test coverage is disabled.
=== RUN   TestDeleteOnlyCrashloopBackoffPods
I0211 16:39:25.850557    7783 restarter.go:111] Starting restarter controller
I0211 16:39:25.850704    7783 restarter.go:113] Starting informer factory.
I0211 16:39:25.851240    7783 restarter.go:119] Waiting for informer caches to sync
I0211 16:39:25.852100    7783 multicontext.go:50] Starting to listen for context start messages
I0211 16:39:25.952125    7783 restarter.go:124] Starting workers
I0211 16:39:25.952338    7783 restarter.go:130] Started workers
I0211 16:39:25.952558    7783 restarter.go:213] Processing endpoint: default/kube-apiserver
I0211 16:39:25.952788    7783 restarter.go:225] Watching for pods in CrashLoopBackOff for a period of 2m0s
I0211 16:39:25.952996    7783 multicontext.go:60] Registering the context for the key: default/kube-apiserver
I0211 16:39:25.953906    7783 restarter.go:322] Deleting pod: pod-c
--- PASS: TestDeleteOnlyCrashloopBackoffPods (2.00s)
    restarter_test.go:246: Starting dep watchdog.
I0211 16:39:27.850768    7783 multicontext.go:54] Received stop signal. Stopping the handling of context messages.
I0211 16:39:27.850787    7783 restarter.go:284] Received stop signal. Stopping watch for pods with selector: garden.sapcloud.io/role in (controlplane)
I0211 16:39:27.850839    7783 restarter.go:132] Shutting down workers
=== RUN   TestDeletePodTransitioningToCrashloopBackoff
==================
WARNING: DATA RACE
Read at 0x00c0000b4910 by goroutine 32:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:142 +0x0
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processEndpoint.func1()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:237 +0x4ed

Previous write at 0x00c0000b4910 by goroutine 16:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:334 +0x0
  github.com/gardener/dependency-watchdog/pkg/multicontext.(*Multicontext).Start()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/multicontext/multicontext.go:56 +0x667

Goroutine 32 (running) created at:
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processEndpoint()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:224 +0x9e5
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processNextWorkItem.func1()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:165 +0x12c
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processNextWorkItem()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:174 +0x6e
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).runWorker()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:141 +0x38
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).runWorker-fm()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:140 +0x41
  k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x61
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x108
  k8s.io/apimachinery/pkg/util/wait.Until()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x5a

Goroutine 16 (finished) created at:
  github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).Run()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:116 +0x27b
  github.com/gardener/dependency-watchdog/pkg/restarter.TestDeleteOnlyCrashloopBackoffPods.func1()
      /tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter_test.go:247 +0x77
==================
panic: send on closed channel

goroutine 57 [running]:
github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processEndpoint.func1(0xc0000e2f20, 0xc0003f6000, 0x16, 0xc0003f6000, 0x7, 0xc0002fc000, 0x1, 0x4)
	/tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:237 +0x4ee
created by github.com/gardener/dependency-watchdog/pkg/restarter.(*Controller).processEndpoint
	/tmp/build/a94a8fe5/pull-request-gardener_dependency-watchdog-pr_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/restarter/restarter.go:224 +0x9e6

Ginkgo ran 1 suite in 45.427064549s
Test Suite Failed`
```

**Comment by amshuman-kr** [comment]
@ialidzhikov Thanks for reporting this. I have [fixed](https://github.com/gardener/dependency-watchdog/pull/16/files#diff-5c2e82b54d23d65b430eed0eb1e1a2ceR47-R49) it in #16.

**Comment by amshuman-kr** [comment]
Fixed in the release [v0.4.1](https://github.com/gardener/dependency-watchdog/releases/tag/v0.4.1).
